File: src/infra/token_sender/stub.rs
use super::{SendError, TokenSender};
use crate::domain::models::{Email, Token};
use async_trait::async_trait;

pub struct StubTokenSender {
    error: bool,
}

impl StubTokenSender {
    #[cfg(test)]
    pub fn new() -> Self {
        Self { error: false }
    }

    #[cfg(test)]
    pub fn with_error(self) -> Self {
        Self {
            error: true,
            ..self
        }
    }
}

#[async_trait]
impl TokenSender for StubTokenSender {
    async fn send(&self, _token: Token, _email: Email) -> Result<(), SendError> {
        if self.error {
            Err(SendError::Unknown("error flag is on"))
        } else {
            Ok(())
        }
    }
}

File: src/infra/tokenSender/stub.ts
import { Email, Token } from "@domain";
import { TokenSender } from "./mod.ts";
import { Result } from "@utils";

export class StubTokenSender implements TokenSender {
    private error: boolean;

    private constructor(error: boolean) {
        this.error = error;
    }

    static new() {
        return new StubTokenSender(false);        
    }

    static withError() {
        return new StubTokenSender(true);
    }

    send(_token: Token, _email: Email): Promise<Result<void, string>> {
        if (this.error) {
            return Promise.resolve(Result.err("error flag is on"));
        } else {
            return Promise.resolve(Result.ok(undefined));
        }
    }
}